gonat_linux.go

```diff
@@ -0,0 +1,337 @@
+	portIndexesByOrigin := s.portIndexes[upFT.IPProto]
+	if portIndexesByOrigin == nil {
+		portIndexesByOrigin = make(map[Addr]int)
+		s.portIndexes[upFT.IPProto] = portIndexesByOrigin
+	}
+
+	upFT.IPProto = downFT.IPProto
+	upFT.Dst = downFT.Dst
+	upFT.Src.IPString = s.opts.IFAddr
+
+	for i := 0; i < numEphemeralPorts; i++ {
+		portIndex := portIndexesByOrigin[downFT.Dst] + 1
+		if portIndex >= numEphemeralPorts {
+			// loop back around to beginning of random sequence
+			portIndex = 0
+		}
+		portIndexesByOrigin[upFT.Dst] = portIndex
+		upFT.Src.Port = s.randomPortSequence[portIndex]
+		err = s.createConntrackEntry(upFT)
+		if err != nil {
+			// this can happen if this 5-tuple is already tracked, ignore and retry
+			continue
+		}
+		return
+	}
+	err = errors.New("Gave up looking for ephemeral port, final error from conntrack: %v", err)
+	return
+}
+
+func (s *server) reapIdleConns() {
+	var connsToClose []*conn
+	for _, c := range s.connsByDownFT {
+		if c.timeSinceLastActive() > s.opts.IdleTimeout {
+			connsToClose = append(connsToClose, c)
+			s.deleteConn(c)
+		}
+	}
+	if len(connsToClose) > 0 {
+		// close conns on a goroutine to avoid tying up main dispatch loop
+		ops.Go(func() {
+			for _, c := range connsToClose {
+				c.Close()
+			}
+		})
+	}
+}
+
+func (s *server) deleteConn(c *conn) {
+	delete(s.connsByDownFT, c.downFT)
+	delete(s.connsByUpFT, c.upFT)
+}
+
+// readFromDownstream reads all IP packets from downstream clients.
+func (s *server) readFromDownstream() error {
+	defer s.Close()
+
+	for {
+		b := s.bufferPool.Get()
+		n, err := s.downstream.Read(b)
+		if err != nil {
+			if err == io.EOF {
+				return err
+			}
+			return errors.New("Unexpected error reading from downstream: %v", err)
+		}
+		raw := b[:n]
+		pkt, err := parseIPPacket(raw)
+		if err != nil {
+			log.Tracef("Error on inbound packet, ignoring: %v", err)
+			s.rejectedPacket()
+			continue
+		}
+		s.fromDownstream <- pkt
+	}
+}
+
+// writeToDownstream writes all IP packets that we're sending back dowstream.
+func (s *server) writeToDownstream() {
+	for pkt := range s.toDownstream {
+		_, err := s.downstream.Write(pkt.Raw)
+		s.bufferPool.Put(pkt.Raw)
+		if err != nil {
+			log.Errorf("Unexpected error writing to downstream: %v", err)
+			return
+		}
+	}
+}
+
+func (s *server) readFromUpstream(socket io.ReadWriteCloser) {
+	defer socket.Close()
+
+	for {
+		b := s.bufferPool.Get()
+		n, err := socket.Read(b)
+		if err != nil {
+			s.rejectedPacket()
+			s.bufferPool.Put(b)
+			if netErr, ok := err.(net.Error); ok && netErr.Temporary() {
+				continue
+			}
+			return
+		}
+		if pkt, err := parseIPPacket(b[:n]); err != nil {
+			log.Tracef("Dropping unparseable packet from upstream: %v", err)
+			s.rejectedPacket()
+			s.bufferPool.Put(b)
+		} else {
+			s.fromUpstream <- pkt
+		}
+	}
+}
+
+func (s *server) Close() error {
+	select {
+	case <-s.close:
+		// already closed
+		return nil
+	default:
+		close(s.close)
+		return nil
```

Comment:
Since this returns before resources are actually closed, something like the following would behave a bit unexpectedly:

```
func main() {
    killSignals := make(chan, os.Signal)
    signal.Notify(killSignals, os.Interrupt, os.Kill)

    s, _ := gonat.NewServer(myDownstreamRW(), nil)
    go s.Serve()
    
    <-killSignals
    log.Fatal(s.Close())
}
```

The runtime would exit when Close returns, but the deferred function in dispatch may not have finished cleaning up yet.  Doesn't seem so bad since it'd just mean some open files and leftover conntrack entries (which will expire).  It'd be unexpected though.  We could fix the situation by calling Serve in the main routine, but that's not obvious from the API.

Reply:
Good point. I'll update Close() to wait until it's actually finished closing.